File: LeetCode/Medium/1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean[] to_delete = new boolean[1001];
    List<TreeNode> ret;
    
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        ret = new ArrayList();
        
        for (int i = 0; i < to_delete.length; ++i) this.to_delete[to_delete[i]] = true;
        
        dfs(root, !this.to_delete[root.val]);
        return ret;
    }
    
    private TreeNode dfs(TreeNode node, boolean isRoot){
        if (node == null) return null;
        
        if (isRoot && !to_delete[node.val]) ret.add(node);
        
        node.left = dfs(node.left, to_delete[node.val]);
        node.right = dfs(node.right, to_delete[node.val]);
        
        if (to_delete[node.val]) return null;
        return node;
    }
}

File: LeetCode/Medium/1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> ret;
    vector<bool> del;
    
    Solution() : del(1001, false) {}
    
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        for (int i = 0; i < to_delete.size(); ++i) this->del[to_delete[i]] = true;
        
        dfs(root, !del[root->val]);
        
        return ret;
    }
    
    TreeNode* dfs(TreeNode* node, bool isRoot){
        if (!node) return nullptr;
        
        if (isRoot && !del[node->val]) ret.push_back(node);
        
        node->left = dfs(node->left, del[node->val]);
        node->right = dfs(node->right, del[node->val]);
        
        if (del[node->val]) return nullptr;
        
        return node;
    }
};